enterprise/models.py

```diff
@@ -1786,9 +1791,59 @@ def get_queryset(self):
         """
         Override to return only those enrollment records for which learner is linked to an enterprise.
         """
+
+        return super().get_queryset().select_related('enterprise_customer_user').filter(
+            enterprise_customer_user__linked=True
+        )
+
+
+class EnterpriseCourseEnrollmentWithAdditionalFieldsManager(models.Manager):
+    """
+    Model manager for `EnterpriseCourseEnrollment`.
+    """
+
+    def get_queryset(self):
+        """
+        Override to return only those enrollment records for which learner is linked to an enterprise.
+        """
+
         return super().get_queryset().select_related('enterprise_customer_user').filter(
             enterprise_customer_user__linked=True
+        ).annotate(**self._get_additional_data_annotations())
+
+    def _get_additional_data_annotations(self):
+        """
+        Return annotations with additional data for the queryset.
+        Additional fields are None in the test environment, where platform models are not available.
+        """
+
+        if not CourseEnrollment or not CourseOverview:
+            return {
+                'enrollment_track': models.Value(None, output_field=models.CharField()),
+                'enrollment_date': models.Value(None, output_field=models.DateTimeField()),
+                'user_email': models.Value(None, output_field=models.EmailField()),
+                'course_start': models.Value(None, output_field=models.DateTimeField()),
+                'course_end': models.Value(None, output_field=models.DateTimeField()),
+            }
+
+        enrollment_subquery = CourseEnrollment.objects.filter(
+            user=models.OuterRef('enterprise_customer_user__user_id'),
+            course_id=models.OuterRef('course_id'),
         )
+        user_subquery = auth.get_user_model().objects.filter(
+            id=models.OuterRef('enterprise_customer_user__user_id'),
+        ).values('email')[:1]
+        course_subquery = CourseOverview.objects.filter(
+            id=models.OuterRef('course_id'),
+        )
+
+        return {
+            'enrollment_track': models.Subquery(enrollment_subquery.values('mode')[:1]),
```

Comment:
@feanil, I've done some tests.

<details>
  <summary>Snippet I used to generate some fake data. Sorry, it's a bit messy. The total number of enterprise enrollments was more than 16000. Other entities are counted in thousands too.</summary>
  
```python
import random
from uuid import uuid4
from django.db.utils import IntegrityError
from common.djangoapps.student.tests.factories import CourseEnrollmentFactory

from openedx.core.djangoapps.content.course_overviews.models import CourseOverview
from enterprise.models import EnterpriseCustomerUser

from enterprise.models import (
    EnterpriseCustomer,
)

from xmodule.modulestore.tests.factories import (
    CourseFactory as ModuleStoreCourseFactory,
    XMODULE_FACTORY_LOCK
)

from common.djangoapps.student.tests.factories import (
    CourseEnrollmentFactory,
    UserFactory,
)
from openedx.core.djangoapps.catalog.tests.factories import (
    CourseFactory,
    CourseRunFactory,
)
from openedx.features.enterprise_support.tests.factories import (
    EnterpriseCourseEnrollmentFactory,
    EnterpriseCustomerFactory,
    EnterpriseCustomerUserFactory,
)

# Attach 10000 enrollments to a single already existing customer
for x in range(10000):
    print(f"First run - {x}")
    try:
        user = UserFactory()
        XMODULE_FACTORY_LOCK.enable()
        modulestore_courses = [ModuleStoreCourseFactory() for _ in range(10)]
        course_runs = [CourseRunFactory(key=str(modulestore_course.id)) for modulestore_course in modulestore_courses]
        courses = [CourseFactory(course_runs=[course_run]) for course_run in course_runs]
        XMODULE_FACTORY_LOCK.disable()
        enterprise_customer = EnterpriseCustomer.objects.get(name="CBA")
        enterprise_customer_user = EnterpriseCustomerUserFactory(
            user_id=user.id,
            enterprise_customer=enterprise_customer
        )
        for modulestore_course in modulestore_courses:
            CourseEnrollmentFactory(
                is_active=True,
                course_id=modulestore_course.id,
                user=user
            )
            EnterpriseCourseEnrollmentFactory(
                course_id=modulestore_course.id,
                enterprise_customer_user=enterprise_customer_user
            )
    except Exception as e:
        print(e)

# Create many more other enterprise users
for x in range(10000):
    print(f"Second run - {x}")
    enterprise_uuid = uuid4()
    try:
        user = UserFactory()
        XMODULE_FACTORY_LOCK.enable()
        modulestore_course = ModuleStoreCourseFactory()
        course_run = CourseRunFactory(key=str(modulestore_course.id))
        course = CourseFactory(course_runs=[course_run])
        XMODULE_FACTORY_LOCK.disable()
        enterprise_customer = EnterpriseCustomerFactory(uuid=enterprise_uuid)
        enterprise_customer_user = EnterpriseCustomerUserFactory(
            user_id=user.id,
            enterprise_customer=enterprise_customer
        )
        CourseEnrollmentFactory(
            is_active=True,
            course_id=modulestore_course.id,
            user=user
        )
        EnterpriseCourseEnrollmentFactory(
            course_id=modulestore_course.id,
            enterprise_customer_user=enterprise_customer_user
        )
    except Exception as e:
        print(e)

# Create many more users that are enrolled into all courses, use random enterprises
courses = list(CourseOverview.objects.all())
enterprise_customer_users = list(EnterpriseCustomerUser.objects.all())
for x in range(1_000_000):
    try:
        print('-' * 100)
        print(x)
        print('-' * 100)
        user = UserFactory()
        for course in [random.choice(courses) for _ in range(20)]:
            CourseEnrollmentFactory(
                is_active=True,
                course_id=str(course.id),
                user=user
            )
            EnterpriseCourseEnrollmentFactory(
                course_id=str(course.id),
                enterprise_customer_user=random.choice(enterprise_customer_users)
            )
    except IntegrityError:
        print("IntegrityError")
        continue
```
  
</details>

<details>
  <summary>Django queryset I used to get an SQL query.</summary>
  
```python
from enterprise.models import EnterpriseCourseEnrollment
qs = EnterpriseCourseEnrollment.with_additional_fields.all()
str(qs.query)
```

</details>

<details>
  <summary>Raw SQL query</summary>
  
```sql
SELECT
  `enterprise_enterprisecourseenrollment`.`id`, 
  `enterprise_enterprisecourseenrollment`.`created`, 
  `enterprise_enterprisecourseenrollment`.`modified`, 
  `enterprise_enterprisecourseenrollment`.`enterprise_customer_user_id`, 
  `enterprise_enterprisecourseenrollment`.`course_id`, 
  `enterprise_enterprisecourseenrollment`.`saved_for_later`, 
  `enterprise_enterprisecourseenrollment`.`source_id`, 
  (
    SELECT 
      U0.`mode` 
    FROM 
      `student_courseenrollment` U0 
    WHERE 
      (
        U0.`course_id` = `enterprise_enterprisecourseenrollment`.`course_id` 
        AND U0.`user_id` = `enterprise_enterprisecustomeruser`.`user_id`
      ) 
    ORDER BY 
      U0.`user_id` ASC, 
      U0.`course_id` ASC 
    LIMIT 
      1
  ) AS `enrollment_track`, 
  (
    SELECT 
      U0.`created` 
    FROM 
      `student_courseenrollment` U0 
    WHERE 
      (
        U0.`course_id` = `enterprise_enterprisecourseenrollment`.`course_id` 
        AND U0.`user_id` = `enterprise_enterprisecustomeruser`.`user_id`
      ) 
    ORDER BY 
      U0.`user_id` ASC, 
      U0.`course_id` ASC 
    LIMIT 
      1
  ) AS `enrollment_date`, 
  (
    SELECT 
      U0.`email` 
    FROM 
      `auth_user` U0 
    WHERE 
      U0.`id` = `enterprise_enterprisecustomeruser`.`user_id` 
    LIMIT 
      1
  ) AS `user_email`, 
  (
    SELECT 
      U0.`start` 
    FROM 
      `course_overviews_courseoverview` U0 
    WHERE 
      U0.`id` = `enterprise_enterprisecourseenrollment`.`course_id` 
    LIMIT 
      1
  ) AS `course_start`, 
  (
    SELECT 
      U0.`end` 
    FROM 
      `course_overviews_courseoverview` U0 
    WHERE 
      U0.`id` = `enterprise_enterprisecourseenrollment`.`course_id` 
    LIMIT 
      1
  ) AS `course_end`, 
  `enterprise_enterprisecustomeruser`.`id`, 
  `enterprise_enterprisecustomeruser`.`created`, 
  `enterprise_enterprisecustomeruser`.`modified`, 
  `enterprise_enterprisecustomeruser`.`enterprise_customer_id`, 
  `enterprise_enterprisecustomeruser`.`user_id`, 
  `enterprise_enterprisecustomeruser`.`active`, 
  `enterprise_enterprisecustomeruser`.`linked`, 
  `enterprise_enterprisecustomeruser`.`is_relinkable`, 
  `enterprise_enterprisecustomeruser`.`invite_key_id`, 
  `enterprise_enterprisecustomeruser`.`should_inactivate_other_customers` 
FROM 
  `enterprise_enterprisecourseenrollment` 
  INNER JOIN `enterprise_enterprisecustomeruser` ON (
    `enterprise_enterprisecourseenrollment`.`enterprise_customer_user_id` = `enterprise_enterprisecustomeruser`.`id`
  ) 
WHERE 
  `enterprise_enterprisecustomeruser`.`linked` 
ORDER BY 
  `enterprise_enterprisecourseenrollment`.`created` ASC
```

</details>

<details>
  <summary>`EXPLAIN FORMAT=json` with `SQL_NO_CACHE`.</summary>

  ```json
{
  "query_block": {
    "select_id": 1,
    "cost_info": {
      "query_cost": "9701.15"
    },
    "ordering_operation": {
      "using_temporary_table": true,
      "using_filesort": true,
      "cost_info": {
        "sort_cost": "4130.80"
      },
      "nested_loop": [
        {
          "table": {
            "table_name": "enterprise_enterprisecustomeruser",
            "access_type": "ALL",
            "possible_keys": [
              "PRIMARY"
            ],
            "rows_examined_per_scan": 2977,
            "rows_produced_per_join": 2679,
            "filtered": "90.00",
            "cost_info": {
              "read_cost": "77.54",
              "eval_cost": "535.86",
              "prefix_cost": "613.40",
              "data_read_per_join": "586K"
            },
            "used_columns": [
              "id",
              "created",
              "modified",
              "user_id",
              "active",
              "linked",
              "enterprise_customer_id",
              "invite_key_id",
              "should_inactivate_other_customers",
              "is_relinkable"
            ],
            "attached_condition": "`edxapp`.`enterprise_enterprisecustomeruser`.`linked`"
          }
        },
        {
          "table": {
            "table_name": "enterprise_enterprisecourseenrollment",
            "access_type": "ref",
            "possible_keys": [
              "enterprise_enterprisecou_enterprise_customer_user_71fe301a_uniq"
            ],
            "key": "enterprise_enterprisecou_enterprise_customer_user_71fe301a_uniq",
            "used_key_parts": [
              "enterprise_customer_user_id"
            ],
            "key_length": "4",
            "ref": [
              "edxapp.enterprise_enterprisecustomeruser.id"
            ],
            "rows_examined_per_scan": 1,
            "rows_produced_per_join": 4130,
            "filtered": "100.00",
            "cost_info": {
              "read_cost": "4130.80",
              "eval_cost": "826.16",
              "prefix_cost": "5570.36",
              "data_read_per_join": "3M"
            },
            "used_columns": [
              "id",
              "created",
              "modified",
              "course_id",
              "enterprise_customer_user_id",
              "source_id",
              "saved_for_later"
            ]
          }
        }
      ],
      "select_list_subqueries": [
        {
          "dependent": true,
          "cacheable": false,
          "query_block": {
            "select_id": 6,
            "cost_info": {
              "query_cost": "1.20"
            },
            "table": {
              "table_name": "U0",
              "access_type": "eq_ref",
              "possible_keys": [
                "PRIMARY"
              ],
              "key": "PRIMARY",
              "used_key_parts": [
                "id"
              ],
              "key_length": "767",
              "ref": [
                "edxapp.enterprise_enterprisecourseenrollment.course_id"
              ],
              "rows_examined_per_scan": 1,
              "rows_produced_per_join": 1,
              "filtered": "100.00",
              "cost_info": {
                "read_cost": "1.00",
                "eval_cost": "0.20",
                "prefix_cost": "1.20",
                "data_read_per_join": "3K"
              },
              "used_columns": [
                "id",
                "end"
              ]
            }
          }
        },
        {
          "dependent": true,
          "cacheable": false,
          "query_block": {
            "select_id": 5,
            "cost_info": {
              "query_cost": "1.20"
            },
            "table": {
              "table_name": "U0",
              "access_type": "eq_ref",
              "possible_keys": [
                "PRIMARY"
              ],
              "key": "PRIMARY",
              "used_key_parts": [
                "id"
              ],
              "key_length": "767",
              "ref": [
                "edxapp.enterprise_enterprisecourseenrollment.course_id"
              ],
              "rows_examined_per_scan": 1,
              "rows_produced_per_join": 1,
              "filtered": "100.00",
              "cost_info": {
                "read_cost": "1.00",
                "eval_cost": "0.20",
                "prefix_cost": "1.20",
                "data_read_per_join": "3K"
              },
              "used_columns": [
                "id",
                "start"
              ]
            }
          }
        },
        {
          "dependent": true,
          "cacheable": false,
          "query_block": {
            "select_id": 4,
            "cost_info": {
              "query_cost": "1.20"
            },
            "table": {
              "table_name": "U0",
              "access_type": "eq_ref",
              "possible_keys": [
                "PRIMARY"
              ],
              "key": "PRIMARY",
              "used_key_parts": [
                "id"
              ],
              "key_length": "4",
              "ref": [
                "edxapp.enterprise_enterprisecustomeruser.user_id"
              ],
              "rows_examined_per_scan": 1,
              "rows_produced_per_join": 1,
              "filtered": "100.00",
              "cost_info": {
                "read_cost": "1.00",
                "eval_cost": "0.20",
                "prefix_cost": "1.20",
                "data_read_per_join": "2K"
              },
              "used_columns": [
                "id",
                "email"
              ],
              "attached_condition": "(`edxapp`.`U0`.`id` = `edxapp`.`enterprise_enterprisecustomeruser`.`user_id`)"
            }
          }
        },
        {
          "dependent": true,
          "cacheable": false,
          "query_block": {
            "select_id": 3,
            "cost_info": {
              "query_cost": "1.20"
            },
            "ordering_operation": {
              "using_filesort": false,
              "table": {
                "table_name": "U0",
                "access_type": "eq_ref",
                "possible_keys": [
                  "student_courseenrollment_user_id_course_id_5d34a47f_uniq",
                  "student_courseenrollment_user_id_4263a8e2",
                  "student_cou_user_id_b19dcd_idx",
                  "student_courseenrollment_course_id_a6f93be8"
                ],
                "key": "student_courseenrollment_user_id_course_id_5d34a47f_uniq",
                "used_key_parts": [
                  "user_id",
                  "course_id"
                ],
                "key_length": "771",
                "ref": [
                  "edxapp.enterprise_enterprisecustomeruser.user_id",
                  "edxapp.enterprise_enterprisecourseenrollment.course_id"
                ],
                "rows_examined_per_scan": 1,
                "rows_produced_per_join": 1,
                "filtered": "100.00",
                "index_condition": "((`edxapp`.`U0`.`user_id` <=> `edxapp`.`enterprise_enterprisecustomeruser`.`user_id`) and (`edxapp`.`U0`.`course_id` <=> `edxapp`.`enterprise_enterprisecourseenrollment`.`course_id`) and (`edxapp`.`U0`.`user_id` = `edxapp`.`enterprise_enterprisecustomeruser`.`user_id`))",
                "cost_info": {
                  "read_cost": "1.00",
                  "eval_cost": "0.20",
                  "prefix_cost": "1.20",
                  "data_read_per_join": "1K"
                },
                "used_columns": [
                  "id",
                  "course_id",
                  "created",
                  "user_id"
                ]
              }
            }
          }
        },
        {
          "dependent": true,
          "cacheable": false,
          "query_block": {
            "select_id": 2,
            "cost_info": {
              "query_cost": "1.20"
            },
            "ordering_operation": {
              "using_filesort": false,
              "table": {
                "table_name": "U0",
                "access_type": "eq_ref",
                "possible_keys": [
                  "student_courseenrollment_user_id_course_id_5d34a47f_uniq",
                  "student_courseenrollment_user_id_4263a8e2",
                  "student_cou_user_id_b19dcd_idx",
                  "student_courseenrollment_course_id_a6f93be8"
                ],
                "key": "student_courseenrollment_user_id_course_id_5d34a47f_uniq",
                "used_key_parts": [
                  "user_id",
                  "course_id"
                ],
                "key_length": "771",
                "ref": [
                  "edxapp.enterprise_enterprisecustomeruser.user_id",
                  "edxapp.enterprise_enterprisecourseenrollment.course_id"
                ],
                "rows_examined_per_scan": 1,
                "rows_produced_per_join": 1,
                "filtered": "100.00",
                "index_condition": "((`edxapp`.`U0`.`user_id` <=> `edxapp`.`enterprise_enterprisecustomeruser`.`user_id`) and (`edxapp`.`U0`.`course_id` <=> `edxapp`.`enterprise_enterprisecourseenrollment`.`course_id`) and (`edxapp`.`U0`.`user_id` = `edxapp`.`enterprise_enterprisecustomeruser`.`user_id`))",
                "cost_info": {
                  "read_cost": "1.00",
                  "eval_cost": "0.20",
                  "prefix_cost": "1.20",
                  "data_read_per_join": "1K"
                },
                "used_columns": [
                  "id",
                  "course_id",
                  "mode",
                  "user_id"
                ]
              }
            }
          }
        }
      ]
    }
  }
}
  ```
</details>

<details>
  <summary>MySQL profiler results.</summary>
  
  
  ```
mysql> show profiles \G;
*************************** 1. row ***************************
Query_ID: 1
Duration: 0.47852300
   Query: SELECT SQL_NO_CACHE `enterprise_enterprisecourseenrollment`.`id`, `enterprise_enterprisecourseenrollment`.`created`, `enterprise_enterprisecourseenrollment`.`modified`, `enterprise_enterprisecourseenrollment`.`enterprise_customer_user_id`, `enterprise_enterprisecourseenrollment`.`course_id`, `enterp
  ```

</details>

<details>
  <summary>Django debug toolbar results for the same query (limited to one organization).</summary>
  
 
![image](https://github.com/openedx/edx-enterprise/assets/18251194/e5941889-fa22-4028-9538-1e404a043809)

  
</details>

So, the query indeed is a bit slow (although, during my tests, the execution time wasn't growing linearly with the number of enrollments) when all records are fetched. However, for non-admin users it's as performant as when the database is almost empty.

I'm not sure if I missed anything @feanil. What do you think? Should I do some other tests?

Reply:
@ormsbee can you take a look at these as well, it looks it has to make a temp table for sorting and that's pretty costly,  and the other major cost seems to be how it's reading the enrollment table.

Reply:
Yeah, the table+filesort is definitely going to be expensive, and it looks like some of the bad reads on `enterprise_enterprisecourseenrollment` might be due to missing some useful indexes?

Questions:

1. Is the performance better if we add a composite `(course_id, enterprise_customer_user)` index to `EnterpriseCourseEnrollment`? (Or even just a simple index on `course_id`?) It's doing a big query on `course_id`, but there's no good index for it. I think it's using the unique constraint that goes the opposite way (`(course_id, enterprise_customer_user)`) and doing a full index scan on that, because that's faster than a full table scan. But an index that leads with `course_id` should be a lot cheaper.
2. Similarly, that table has an implicit ordering of `created`, but seems to have no index on that field. Do we really need to order at all? Can we add on there? Is an ordering by `created` equivalent to ordering by the primary key, and can we just do that for the purposes of this query?
3. Does this make use of [`read_replica_or_default`](https://github.com/openedx/edx-django-utils/blame/2c975229cbbbfc07f3320b72cbb91f208957f2f5/edx_django_utils/db/read_replica.py#L43)? If not, can we make it use it?

Adding new indexes may be slow, so we'd have to communicate it so as not to be disruptive.

Reply:
@feanil, @ormsbee, thank you for checking this. I'll try adding proposed indexes and will let you know how performance changes.